Add execution order groups (#23473)
idaholab/moose: framework/doc/content/syntax/UserObjects/index.md

4. `threadJoin` and `finalize` are called in the following order

   1. for objects derived `SideUserObject`
   2. for objects derived `InternalSideUserObject`
   3. for objects derived `InterfaceUserObject`
   4. for objects derived `ElementUserObject`
   5. for objects derived `DomainUserObject`

5. `initialize()` is called for objects derived from `NodalUserObject`.

6. `NodalUserObject` are looped over all local nodes and executed on each node (here the order is determined through dependency resolution).

7. `threadJoin` and `finalize` are called for `NodalUserObject`s.

8. `initialize()` is called for objects derived from `ThreadedGeneralUserObject`.

9. `execute()` is called for objects derived from `ThreadedGeneralUserObject` in a threaded way.

10. `threadJoin` and `finalize` ar called for `ThreadedGeneralUserObject`s.

12. `initialize`, `execute`, and `finalize` are called for each `GeneralUserObject`.

> For additional control over the user object execution order very user object has a `execution_order_group`

Reviewer: `very` => `every`.

Author: Can we use `!param`? `UserObject` is not a registered class, so what syntax would I link this to?

Reviewer: Oh, then probably not. We cannot refer to a specific user object here.

parameter of type integer. This parameter can be used to force multiple execution rounds of the
above order. All objects with the same `execution_order_group` parameter value are executed in the

> same round, and grups with a lower number are executed first. The default `execution_order_group`

Reviewer: grups => groups.

> is 0 (zero). Negative values can be specified to force UO execution _before_ the default group, and

Reviewer: UO to UO (user object) if it does not appear before in this page.

- +DomainUserObject+: this object is capable of executing all the operations of
  a +ElementUserObject+, +InternalSideUserObject+, +SideUserObject+ and +InterfaceUserObject+.

# Execution order

> 

Author: link to where?

> 

Reviewer 2: the poorly named SetupInterface

3. All active local elements are iterated over and objects derived from `ElementUserObject`,
   `SideUserObject`, `InternalSideUserObject` `InterfaceUserObject`, and  `DomainUserObject` are
   executed on elements, boundaries, internal sides, and on subdomain changes (here the order
   within each type group is determined through dependency resolution).

4. `threadJoin` and `finalize` are called in the following order

   1. for objects derived `SideUserObject`
   2. for objects derived `InternalSideUserObject`
   3. for objects derived `InterfaceUserObject`
   4. for objects derived `ElementUserObject`
   5. for objects derived `DomainUserObject`

5. `initialize` is called for objects derived from `NodalUserObject`.

6. `NodalUserObject` are looped over all local nodes and executed on each node (here the order
   is determined through dependency resolution).

7. `threadJoin` and `finalize` are called for `NodalUserObject`s.

8. `initialize` is called for objects derived from `ThreadedGeneralUserObject`.

9. `execute` is called for objects derived from `ThreadedGeneralUserObject` in a threaded way.

10. `threadJoin` and `finalize` ar called for `ThreadedGeneralUserObject`s.

> 12. `initialize`, `execute`, and `finalize` are called for each `GeneralUserObject`.

Author: in that order

> 12. `initialize`, `execute`, and `finalize` are called for each `GeneralUserObject`.

Reviewer 2: lolol

ordering of GeneralUserObject.
Is there dependency resolution? Is it alphabetical?

> 12. `initialize`, `execute`, and `finalize` are called for each `GeneralUserObject`.

Author: The list comes from a warehouse query, so I'm assuming there is dependency resolution at least within the GUOs. I'll check and amend.

   1. for objects derived from `ElementUserObject`,  `SideUserObject`, `InternalSideUserObject`
      `InterfaceUserObject`, and  `DomainUserObject`.
   2. for objects derived from `NodalUserObject`.
   3. for objects derived from `ThreadedGeneralUserObject`.
   4. for objects derived from `GeneralUserObject`.

2. `initialize` is called for objects derived from `ElementUserObject`,  `SideUserObject`,
   `InternalSideUserObject` `InterfaceUserObject`, and  `DomainUserObject` in that order.

3. All active local elements are iterated over and objects derived from `ElementUserObject`,
   `SideUserObject`, `InternalSideUserObject` `InterfaceUserObject`, and  `DomainUserObject` are
   executed on elements, boundaries attached to the element, internal sides between elements, and on subdomain changes, in that order.
   The order within each type group is determined through dependency resolution.

4. `threadJoin` and `finalize` are called in the following order

   1. for objects derived from `SideUserObject`
   2. for objects derived from `InternalSideUserObject`
   3. for objects derived from `InterfaceUserObject`
   4. for objects derived from `ElementUserObject`

>    5. for objects derived from `DomainUserObject`

Reviewer 3: So this inconsistent ordering is because of Griffin? Is the plan to merge this, use the execution order feature in Griffin where needed, and then fix this ordering to be consistent?

Author: Um, I don't have any plans to fix griffin. Why is this ordering inconsistent? I also suspect that after years of living with this ordering other apps rely on this ordering. We can certainly explore changing that order, but that'll be a different PR.

Reviewer 4: Yeah I don't think we can get around this default and I'm fine with that

Reviewer: If I recall correctly, we used to have `ElementUserObject` as a main object that gathers info from `SideUserObject`s in Griffin. I am not exactly sure what you mean inconsistency, maybe the ordering in setups? But I think after introducing `DomainUserObject`, we got rid of this dependency. Either way, I do not think it is wise to change something in this PR which is introducing a new feature for now.

Reviewer 3: Just to clarify, my question is not a requirement for this PR - it should be merged regardless.

By "inconsistent" I meant that in all of the other steps, the UO ordering is {element, side, internal side, interface, domain}. There was a comment in the code:
```
// There is one instance in rattlesnake where an elemental user object's finalize depends
// on a side user object having been finalized first :-(
```
I assumed that this was the reason.

Again, I'm not asking for this change in this PR, and like you said, there's a decent chance that other apps are already relying on this ordering, so it's fine to keep. It was just a minor confusion.

Reviewer: Oh, that is a very old comment. We possibly should make sure all these mesh related user object types are independent by issuing errors if such a UO tries to get another UO. Users can probably use DomainUserObject or GeneralUserObject to resolve the dependency. If we want some implicit ordering across these mesh UOs, ElementUserObject getting executed last seems a good choice to me.

Reviewer 3: > issuing errors if such a UO tries to get another UO

I bet some apps will get caught by this, but I think there should be ways to avoid hitting this.

Anyway, I added https://github.com/idaholab/moose/issues/23615 for if we ever want to explore this, and I'll resolve this conversation.

6. `NodalUserObject` are looped over all local nodes and executed on each node.
   On each node, the order is determined through dependency resolution.

7. `threadJoin` and `finalize` are called for `NodalUserObject`s.

8. `initialize` is called for objects derived from `ThreadedGeneralUserObject`.

9. `execute` is called for objects derived from `ThreadedGeneralUserObject` in a threaded way.

10. `threadJoin` and `finalize` ar called for `ThreadedGeneralUserObject`s.

12. `initialize`, `execute`, and `finalize` are called in that order for each `GeneralUserObject` (which in turn are ordered through dependency resolution within the set of applicable `GeneralUserObject`s).

For additional control over the user object execution order every user object has a `execution_order_group`
parameter of type integer. This parameter can be used to force multiple execution rounds of the
above order, so it effectively supersedes all the ordering described above. All objects with the same `execution_order_group` parameter value are executed in the
same round, and groups with a lower number are executed first. The default `execution_order_group`

> is 0 (zero). Negative values can be specified to force user object execution _before_ the default group, and
> positive values can be uses to create execution groups that run _after_ the default group. Execution

Reviewer 3: ```suggestion
is 0 (zero). Negative values can be specified to force user object execution *before* the default group, and
positive values can be uses to create execution groups that run *after* the default group. Execution
```

- +DomainUserObject+: this object is capable of executing all the operations of
  a +ElementUserObject+, +InternalSideUserObject+, +SideUserObject+ and +InterfaceUserObject+.

# Execution order

> With in an execution stage set by [`execute_on`](SetupInterface.md) user objects are executed in

Reviewer 2: ```suggestion
Within an execution stage set by the [`execute_on`](SetupInterface.md) parameter, user objects are executed in
```

the following order:

1. `residualSetup` / `jacobianSetup`

   If the current `execute_on` flag is either `EXEC_LINEAR` or `EXEC_NONLINEAR` the `residualSetup`
   and `jacobianSetup` are called respectively in the following order

   1. for objects derived from `ElementUserObject`,  `SideUserObject`, `InternalSideUserObject`
      `InterfaceUserObject`, and  `DomainUserObject`.
   2. for objects derived from `NodalUserObject`.
   3. for objects derived from `ThreadedGeneralUserObject`.
   4. for objects derived from `GeneralUserObject`.

2. `initialize` is called for objects derived from `ElementUserObject`,  `SideUserObject`,
   `InternalSideUserObject` `InterfaceUserObject`, and  `DomainUserObject` in that order.

3. All active local elements are iterated over and objects derived from `ElementUserObject`,
   `SideUserObject`, `InternalSideUserObject` `InterfaceUserObject`, and  `DomainUserObject` are
   executed on elements, boundaries attached to the element, internal sides between elements, and on subdomain changes, in that order.

>    The order
>    within each type group is determined through dependency resolution.

Reviewer 2: ```suggestion
   The order within each type group is determined through dependency resolution.
```
